// validor/dist/errors/ValidationError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidationError = void 0;
class ValidationError extends Error {
    constructor(issues) {
        super(ValidationError.buildMessage(issues));
        this.issues = issues;
        this.name = "ValidationError";
        Error.captureStackTrace?.(this, ValidationError);
    }
    static buildMessage(issues) {
        if (issues.length === 0)
            return "Validation failed";
        const first = issues[0].message;
        const remaining = issues.length - 1;
        return remaining > 0
            ? `${first} (and ${remaining} more error${remaining > 1 ? "s" : ""})`
            : first;
    }
    /**
     * Format issues as an array, optionally applying locale messages
     * @param localeMessages - Optional mapping of error codes to localized messages
     */
    format(localeMessages) {
        return this.issues.map(({ path = [], message, code }) => ({
            path,
            message: localeMessages?.[code] ?? message,
            code,
        }));
    }
    /**
     * Format errors grouped by field path
     */
    formatErrors() {
        const errors = {};
        for (const { path, message } of this.issues) {
            const key = path.join(".");
            if (!key)
                continue;
            (errors[key] ?? (errors[key] = [])).push(message);
        }
        return { message: this.message, errors };
    }
    /**
     * @deprecated Use format(localeMessages) instead
     */
    formatWithLocale(localeMessages) {
        return this.format(localeMessages);
    }
}
exports.ValidationError = ValidationError;
