// validor/dist/core/BaseSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseSchema = void 0;
const ValidationError_1 = require("../errors/ValidationError");
const messages_1 = require("../i18n/messages");
class BaseSchema {
    constructor() {
        this.isRequiredFlag = false;
        this.isNullable = false;
        this.path = [];
    }
    /**
     * Set the required flag for the schema
     * @param message - The message to display if the value is required
     * @returns The schema instance
     *
     * @example
     * const schema = string().required();
     * const schema = string().required("This field is required");
     */
    required(message) {
        this.isRequiredFlag = true;
        this.requiredMessage = message;
        return this;
    }
    /**
     * Set the optional flag for the schema
     * @returns The schema instance
     *
     * @example
     * const schema = string().optional();
     */
    optional() {
        this.isRequiredFlag = false;
        this.requiredMessage = undefined;
        return this;
    }
    /**
     * Set the nullable flag for the schema
     * @returns The schema instance
     *
     * @example
     * const schema = string().nullable();
     */
    nullable() {
        this.isNullable = true;
        return this;
    }
    /**
     * Set the default value for the schema
     * @param value - The default value to use
     * @returns The schema instance
     *
     * @example
     * const schema = string().default("hello");
     */
    default(value) {
        this.defaultValue = value;
        return this;
    }
    /**
     * Transform the value of the schema
     * @param fn - The transformation function
     * @returns The schema instance
     *
     * @example
     * const schema = string().transform((value) => value.toUpperCase());
     */
    transform(fn) {
        this.transformFn = fn;
        return this;
    }
    /**
     * Parse the value of the schema
     * @param value - The value to parse
     * @param path - The path to the value
     * @returns The parsed value
     *
     * @example
     * const result = schema.safeParse("hello");
     * console.log(result.success); // true
     * console.log(result.data); // "HELLO"
     */
    safeParse(value, path = []) {
        this.path = path;
        try {
            const result = this.processValue(value, path);
            return { success: true, data: result };
        }
        catch (err) {
            const error = err instanceof ValidationError_1.ValidationError
                ? err
                : new ValidationError_1.ValidationError([
                    {
                        path,
                        message: err instanceof Error ? err.message : "Validation failed",
                        code: "invalid",
                    },
                ]);
            return { success: false, error };
        }
    }
    /**
     * Validate the value of the schema
     * @param value - The value to validate
     * @param path - The path to the value
     * @returns The validated value
     *
     * @example
     * const result = await schema.validate("hello");
     */
    async validate(value, path = []) {
        const result = this.safeParse(value, path);
        if (result.success === false) {
            throw result.error;
        }
        return result.data;
    }
    // All value processing logic in one place
    processValue(value, path) {
        if (value === undefined && this.defaultValue !== undefined) {
            value =
                typeof this.defaultValue === "function"
                    ? this.defaultValue()
                    : this.defaultValue;
        }
        // Handle null
        if (value === null) {
            if (this.isNullable)
                return null;
            if (this.isRequiredFlag)
                this.throwRequired(path);
            return null;
        }
        // Handle undefined
        if (value === undefined) {
            if (this.isRequiredFlag)
                this.throwRequired(path);
            return undefined;
        }
        // Parse and transform
        const parsed = this.parse(value, path);
        return this.transformFn ? this.transformFn(parsed) : parsed;
    }
    throwRequired(path) {
        throw new ValidationError_1.ValidationError([
            {
                path,
                message: this.requiredMessage || (0, messages_1.getMessage)("required", path),
                code: "required",
            },
        ]);
    }
    /**
     * Get the resolved default value if set
     */
    getDefaultValue() {
        if (this.defaultValue === undefined)
            return undefined;
        return typeof this.defaultValue === "function"
            ? this.defaultValue()
            : this.defaultValue;
    }
    /**
     * Throw a type validation error
     */
    throwTypeError(path, expectedType) {
        throw new ValidationError_1.ValidationError([
            {
                path,
                message: (0, messages_1.getMessage)("invalid_type", path, expectedType ? { expected: expectedType } : undefined),
                code: "invalid_type",
            },
        ]);
    }
    /**
     * Handle null/undefined values with defaults and required checks
     * Returns {handled: true, value: T} if value was handled, or {handled: false} to continue parsing
     */
    handleNullUndefined(value, path) {
        // Apply default if value is undefined
        if (value === undefined) {
            const defaultVal = this.getDefaultValue();
            if (defaultVal !== undefined) {
                return { handled: true, value: defaultVal };
            }
            if (this.isRequiredFlag) {
                this.throwRequired(path);
            }
            return { handled: true, value: undefined };
        }
        // Handle null values
        if (value === null) {
            const defaultVal = this.getDefaultValue();
            if (defaultVal !== undefined) {
                return { handled: true, value: defaultVal };
            }
            if (this.isNullable) {
                return { handled: true, value: null };
            }
            if (this.isRequiredFlag) {
                this.throwRequired(path);
            }
            return { handled: true, value: null };
        }
        // Signal to continue with normal parsing
        return { handled: false };
    }
}
exports.BaseSchema = BaseSchema;
// Protected helper methods for child schemas
// Sentinel object to signal that parsing should continue
BaseSchema.CONTINUE_PARSING = Symbol("CONTINUE_PARSING");
