// validor/dist/schemas/ArraySchema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArraySchema = void 0;
const BaseSchema_1 = require("../core/BaseSchema");
const ValidationError_1 = require("../errors/ValidationError");
const messages_1 = require("../i18n/messages");
/**
 * Schema class for validating array values.
 * Supports element validation, length constraints, and array-specific checks.
 *
 * @template T - The type of elements in the array
 *
 * @example
 * const schema = new ArraySchema(new StringSchema());
 * schema.parse(["a", "b", "c"]); // ["a", "b", "c"]
 *
 * @example
 * // With constraints
 * const schema = new ArraySchema(new NumberSchema())
 *   .min(1)
 *   .max(10)
 *   .unique();
 */
class ArraySchema extends BaseSchema_1.BaseSchema {
    /**
     * Creates a new ArraySchema instance.
     *
     * @public
     * @param elementSchema - Optional schema to validate each array element
     *
     * @example
     * const stringArraySchema = new ArraySchema(new StringSchema());
     * const numberArraySchema = new ArraySchema(new NumberSchema().min(0));
     * const anyArraySchema = new ArraySchema(); // No element validation
     */
    constructor(elementSchema) {
        super();
        /**
         * List of validation checks to run against the array.
         * Each check contains a test function, error code, and optional custom message.
         * @private
         */
        this.checks = [];
        /**
         * Flag indicating whether to coerce non-array values to single-element arrays.
         * @private
         */
        this.coerceFlag = false;
        this.elementSchema = elementSchema;
    }
    /**
     * Parses and validates the given value as an array.
     *
     * @public
     * @param value - The value to validate
     * @param path - The field path for error messages
     * @returns The validated array with all elements parsed
     * @throws {ValidationError} If validation fails
     *
     * @example
     * const schema = new ArraySchema(new StringSchema());
     * schema.parse(["a", "b"]); // ["a", "b"]
     * schema.parse("not-array"); // throws ValidationError
     */
    parse(value, path = []) {
        // Handle null/undefined
        if (value == null) {
            if (this.defaultValue !== undefined) {
                return typeof this.defaultValue === "function"
                    ? this.defaultValue()
                    : this.defaultValue;
            }
            if (this.isRequiredFlag) {
                throw new ValidationError_1.ValidationError([
                    {
                        path,
                        message: (0, messages_1.getMessage)("required", path),
                        code: "required",
                    },
                ]);
            }
            return value;
        }
        // Apply coercion if enabled
        let arrayValue = value;
        if (this.coerceFlag && !Array.isArray(value)) {
            arrayValue = [value];
        }
        // Type check
        if (!Array.isArray(arrayValue)) {
            throw new ValidationError_1.ValidationError([
                {
                    path,
                    message: (0, messages_1.getMessage)("invalid_type", path, { expected: "array" }),
                    code: "invalid_type",
                },
            ]);
        }
        // Validate each element if elementSchema is provided
        const issues = [];
        let parsedElements;
        if (this.elementSchema) {
            parsedElements = [];
            for (let i = 0; i < arrayValue.length; i++) {
                try {
                    const elementPath = [...path, i.toString()];
                    const parsedElement = this.elementSchema.parse(arrayValue[i], elementPath);
                    parsedElements.push(parsedElement);
                }
                catch (error) {
                    if (error instanceof ValidationError_1.ValidationError) {
                        issues.push(...error.issues);
                    }
                    else {
                        throw error;
                    }
                }
            }
            if (issues.length > 0) {
                throw new ValidationError_1.ValidationError(issues);
            }
        }
        else {
            // No element validation, use array as-is
            parsedElements = arrayValue;
        }
        // Run array-level checks
        const arrayIssues = this.checks
            .filter(({ test }) => !test(parsedElements))
            .map(({ code, message, params }) => ({
            path,
            message: message || (0, messages_1.getMessage)(code, path, params),
            code,
        }));
        if (arrayIssues.length > 0) {
            throw new ValidationError_1.ValidationError(arrayIssues);
        }
        // Apply transform if exists
        return this.transformFn ? this.transformFn(parsedElements) : parsedElements;
    }
    /**
     * Adds a new validation check to the schema.
     *
     * @private
     * @param test - The test function that returns true if validation passes
     * @param code - The error code for i18n message lookup
     * @param message - Optional custom error message
     * @param params - Optional parameters to include in error message
     * @returns this for method chaining
     */
    addCheck(test, code, message, params) {
        this.checks.push({ test, code, message, params });
        return this;
    }
    /**
     * Enables coercion mode to wrap non-array values in an array.
     * Single values will be converted to single-element arrays.
     *
     * @public
     * @returns this for method chaining
     *
     * @example
     * const schema = new ArraySchema(new StringSchema()).coerce();
     * schema.parse("hello"); // ["hello"]
     * schema.parse(["a", "b"]); // ["a", "b"]
     */
    coerce() {
        this.coerceFlag = true;
        return this;
    }
    /**
     * Adds a minimum length validation check.
     *
     * @public
     * @param n - The minimum number of elements required
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * const schema = new ArraySchema(new StringSchema()).min(2);
     * schema.parse(["a"]);      // throws ValidationError
     * schema.parse(["a", "b"]); // ["a", "b"]
     */
    min(n, message) {
        return this.addCheck((val) => val.length >= n, "array_too_small", message, {
            min: n,
        });
    }
    /**
     * Adds a maximum length validation check.
     *
     * @public
     * @param n - The maximum number of elements allowed
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * const schema = new ArraySchema(new StringSchema()).max(3);
     * schema.parse(["a", "b"]);           // ["a", "b"]
     * schema.parse(["a", "b", "c", "d"]); // throws ValidationError
     */
    max(n, message) {
        return this.addCheck((val) => val.length <= n, "array_too_big", message, {
            max: n,
        });
    }
    /**
     * Adds an exact length validation check.
     *
     * @public
     * @param n - The exact number of elements required
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * const schema = new ArraySchema(new StringSchema()).length(3);
     * schema.parse(["a", "b", "c"]); // ["a", "b", "c"]
     * schema.parse(["a", "b"]);      // throws ValidationError
     */
    length(n, message) {
        return this.addCheck((val) => val.length === n, "array_invalid_length", message, {
            length: n,
        });
    }
    /**
     * Adds a non-empty array validation check.
     * Ensures the array has at least one element.
     *
     * @public
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * const schema = new ArraySchema(new StringSchema()).nonempty();
     * schema.parse(["a"]); // ["a"]
     * schema.parse([]);    // throws ValidationError
     */
    nonempty(message) {
        return this.addCheck((val) => val.length > 0, "array_empty", message);
    }
    /**
     * Adds a uniqueness validation check.
     * Ensures all elements in the array are unique (using strict equality).
     *
     * @public
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * const schema = new ArraySchema(new NumberSchema()).unique();
     * schema.parse([1, 2, 3]);    // [1, 2, 3]
     * schema.parse([1, 2, 2, 3]); // throws ValidationError
     */
    unique(message) {
        return this.addCheck((val) => new Set(val).size === val.length, "array_not_unique", message);
    }
    /**
     * Adds a uniqueness validation check using a custom key extractor.
     * Useful for checking uniqueness based on object properties.
     *
     * @public
     * @param keyFn - Function to extract the comparison key from each element
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * const schema = new ArraySchema(new ObjectSchema({...}))
     *   .uniqueBy((item) => item.id);
     * schema.parse([{id: 1}, {id: 2}]); // OK
     * schema.parse([{id: 1}, {id: 1}]); // throws ValidationError
     */
    uniqueBy(keyFn, message) {
        return this.addCheck((val) => {
            const keys = val.map(keyFn);
            return new Set(keys).size === keys.length;
        }, "array_not_unique", message);
    }
    /**
     * Adds a validation check to ensure the array contains a specific value.
     *
     * @public
     * @param value - The value that must be present in the array
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * const schema = new ArraySchema(new StringSchema()).contains("admin");
     * schema.parse(["user", "admin"]); // ["user", "admin"]
     * schema.parse(["user", "guest"]); // throws ValidationError
     */
    contains(value, message) {
        return this.addCheck((val) => val.includes(value), "array_missing_element", message, {
            value,
        });
    }
    /**
     * Adds a validation check using a custom predicate.
     * At least one element must satisfy the predicate.
     *
     * @public
     * @param predicate - Function that returns true if element satisfies condition
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * const schema = new ArraySchema(new NumberSchema())
     *   .some((n) => n > 10);
     * schema.parse([1, 5, 15]); // [1, 5, 15]
     * schema.parse([1, 5, 9]);  // throws ValidationError
     */
    some(predicate, message) {
        return this.addCheck((val) => val.some(predicate), "array_none_match", message);
    }
    /**
     * Adds a validation check using a custom predicate.
     * All elements must satisfy the predicate.
     *
     * @public
     * @param predicate - Function that returns true if element satisfies condition
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * const schema = new ArraySchema(new NumberSchema())
     *   .every((n) => n > 0);
     * schema.parse([1, 2, 3]);  // [1, 2, 3]
     * schema.parse([1, -2, 3]); // throws ValidationError
     */
    every(predicate, message) {
        return this.addCheck((val) => val.every(predicate), "array_not_all_match", message);
    }
    /**
     * Adds a validation check for array length range.
     * Ensures the array length is between min and max (inclusive).
     *
     * @public
     * @param min - Minimum number of elements
     * @param max - Maximum number of elements
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * const schema = new ArraySchema(new StringSchema()).between(2, 5);
     * schema.parse(["a", "b", "c"]); // ["a", "b", "c"]
     * schema.parse(["a"]);          // throws ValidationError
     */
    between(min, max, message) {
        return this.addCheck((val) => val.length >= min && val.length <= max, "array_length_out_of_range", message, { min, max });
    }
    /**
     * Adds a validation check to ensure no duplicate values based on JSON serialization.
     * Useful for arrays of objects where deep equality is needed.
     *
     * @public
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * const schema = new ArraySchema(new ObjectSchema({...})).uniqueDeep();
     * schema.parse([{a: 1}, {a: 2}]); // OK
     * schema.parse([{a: 1}, {a: 1}]); // throws ValidationError
     */
    uniqueDeep(message) {
        return this.addCheck((val) => {
            const serialized = val.map((item) => JSON.stringify(item));
            return new Set(serialized).size === serialized.length;
        }, "array_not_unique", message);
    }
    /**
     * Adds a validation check to ensure the array is sorted in ascending order.
     *
     * @public
     * @param compareFn - Optional comparison function (defaults to standard comparison)
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * const schema = new ArraySchema(new NumberSchema()).sorted();
     * schema.parse([1, 2, 3]); // [1, 2, 3]
     * schema.parse([1, 3, 2]); // throws ValidationError
     */
    sorted(compareFn, message) {
        return this.addCheck((val) => {
            if (val.length <= 1)
                return true;
            const compare = compareFn || ((a, b) => (a < b ? -1 : a > b ? 1 : 0));
            for (let i = 1; i < val.length; i++) {
                if (compare(val[i - 1], val[i]) > 0)
                    return false;
            }
            return true;
        }, "array_not_sorted", message);
    }
    /**
     * Adds a validation check to ensure the array is sorted in descending order.
     *
     * @public
     * @param compareFn - Optional comparison function (defaults to standard comparison)
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * const schema = new ArraySchema(new NumberSchema()).sortedDesc();
     * schema.parse([3, 2, 1]); // [3, 2, 1]
     * schema.parse([1, 3, 2]); // throws ValidationError
     */
    sortedDesc(compareFn, message) {
        return this.addCheck((val) => {
            if (val.length <= 1)
                return true;
            const compare = compareFn || ((a, b) => (a < b ? -1 : a > b ? 1 : 0));
            for (let i = 1; i < val.length; i++) {
                if (compare(val[i - 1], val[i]) < 0)
                    return false;
            }
            return true;
        }, "array_not_sorted_desc", message);
    }
    /**
     * Adds a validation check to ensure all elements are of the same type.
     * Checks the typeof each element.
     *
     * @public
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * const schema = new ArraySchema(new AnySchema()).homogeneous();
     * schema.parse([1, 2, 3]);     // [1, 2, 3]
     * schema.parse([1, "2", 3]);   // throws ValidationError
     */
    homogeneous(message) {
        return this.addCheck((val) => {
            if (val.length <= 1)
                return true;
            const firstType = typeof val[0];
            return val.every((item) => typeof item === firstType);
        }, "array_not_homogeneous", message);
    }
    /**
     * Adds a validation check to ensure the array does not contain a specific value.
     *
     * @public
     * @param value - The value that must not be present in the array
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * const schema = new ArraySchema(new StringSchema()).excludes("banned");
     * schema.parse(["allowed", "ok"]);   // ["allowed", "ok"]
     * schema.parse(["allowed", "banned"]); // throws ValidationError
     */
    excludes(value, message) {
        return this.addCheck((val) => !val.includes(value), "array_contains_forbidden", message, {
            value,
        });
    }
    /**
     * Adds a custom validation check to the array.
     *
     * @public
     * @param validator - Custom validation function
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * const schema = new ArraySchema(new NumberSchema())
     *   .refine((arr) => arr.reduce((a, b) => a + b, 0) <= 100, "Sum must be <= 100");
     * schema.parse([10, 20, 30]); // [10, 20, 30]
     * schema.parse([50, 60, 70]); // throws ValidationError
     */
    refine(validator, message) {
        return this.addCheck(validator, "array_custom_validation", message);
    }
}
exports.ArraySchema = ArraySchema;
