// validor/dist/schemas/BooleanSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BooleanSchema = void 0;
const BaseSchema_1 = require("../core/BaseSchema");
const ValidationError_1 = require("../errors/ValidationError");
const messages_1 = require("../i18n/messages");
/**
 * Schema class for validating boolean values.
 * Supports strict mode and coercion from common truthy/falsy values.
 *
 * @example
 * const schema = new BooleanSchema();
 * schema.parse(true);  // true
 * schema.parse(false); // false
 *
 * @example
 * // With coercion enabled
 * const schema = new BooleanSchema().coerce();
 * schema.parse("true"); // true
 * schema.parse(1);      // true
 * schema.parse("false"); // false
 * schema.parse(0);       // false
 */
class BooleanSchema extends BaseSchema_1.BaseSchema {
    constructor() {
        super(...arguments);
        /**
         * Flag indicating whether to coerce string/number values to boolean.
         * When enabled, values like "true", "false", 1, 0 are converted to boolean.
         * @private
         */
        this.coerceFlag = false;
        /**
         * Constraint flags for literal value validation
         * @private
         */
        this.mustBeTrue = false;
        this.mustBeFalse = false;
        /**
         * List of values that should be coerced to true.
         * @private
         */
        this.truthyValues = new Set([
            true,
            "true",
            "1",
            1,
        ]);
        /**
         * List of values that should be coerced to false.
         * @private
         */
        this.falsyValues = new Set([
            false,
            "false",
            "0",
            0,
        ]);
    }
    /**
     * Parses and validates the given value as a boolean.
     *
     * @public
     * @param value - The value to validate
     * @param path - The field path for error messages
     * @returns The validated boolean value
     * @throws {ValidationError} If validation fails
     *
     * @example
     * const schema = new BooleanSchema();
     * schema.parse(true);    // true
     * schema.parse("true");  // throws ValidationError (unless coerce is enabled)
     */
    parse(value, path = []) {
        // Handle null/undefined with defaults
        const handleResult = this.handleNullUndefined(value, path);
        if (handleResult.handled) {
            return handleResult.value;
        }
        // Apply coercion if enabled
        let result = value;
        if (this.coerceFlag) {
            result = this.coerceValue(value);
        }
        // Type check
        if (typeof result !== "boolean") {
            this.throwTypeError(path, "boolean");
        }
        // Check literal constraints
        if (this.mustBeTrue && result !== true) {
            throw new ValidationError_1.ValidationError([
                {
                    path,
                    message: this.trueLiteralMessage ||
                        (0, messages_1.getMessage)("invalid_literal", path, { expected: "true" }),
                    code: "invalid_literal",
                },
            ]);
        }
        if (this.mustBeFalse && result !== false) {
            throw new ValidationError_1.ValidationError([
                {
                    path,
                    message: this.falseLiteralMessage ||
                        (0, messages_1.getMessage)("invalid_literal", path, { expected: "false" }),
                    code: "invalid_literal",
                },
            ]);
        }
        // Apply transform if exists
        return this.transformFn ? this.transformFn(result) : result;
    }
    /**
     * Coerces common truthy/falsy values to boolean.
     *
     * @private
     * @param value - The value to coerce
     * @returns The coerced boolean or original value if not coercible
     */
    coerceValue(value) {
        // Normalize string values to lowercase for comparison
        const normalizedValue = typeof value === "string" ? value.toLowerCase().trim() : value;
        if (this.truthyValues.has(normalizedValue)) {
            return true;
        }
        if (this.falsyValues.has(normalizedValue)) {
            return false;
        }
        return value;
    }
    /**
     * Enables coercion mode to convert common truthy/falsy values to boolean.
     * Accepted truthy values: true, "true", "1", 1, "yes", "on"
     * Accepted falsy values: false, "false", "0", 0, "no", "off"
     *
     * @public
     * @returns this for method chaining
     *
     * @example
     * const schema = new BooleanSchema().coerce();
     * schema.parse("true");  // true
     * schema.parse("yes");   // true
     * schema.parse(1);       // true
     * schema.parse("false"); // false
     * schema.parse("no");    // false
     * schema.parse(0);       // false
     */
    coerce() {
        this.coerceFlag = true;
        return this;
    }
    /**
     * Validates that the value is strictly true.
     *
     * @public
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * const schema = new BooleanSchema().true();
     * schema.parse(true);  // true
     * schema.parse(false); // throws ValidationError
     */
    true(message) {
        this.mustBeTrue = true;
        this.trueLiteralMessage = message;
        return this;
    }
    /**
     * Validates that the value is strictly false.
     *
     * @public
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * const schema = new BooleanSchema().false();
     * schema.parse(false); // false
     * schema.parse(true);  // throws ValidationError
     */
    false(message) {
        this.mustBeFalse = true;
        this.falseLiteralMessage = message;
        return this;
    }
    /**
     * Adds custom truthy values for coercion.
     * These values will be converted to true when coerce() is enabled.
     *
     * @public
     * @param values - Array of values to treat as truthy
     * @returns this for method chaining
     *
     * @example
     * const schema = new BooleanSchema()
     *   .coerce()
     *   .addTruthyValues(["да", "hai", "oui"]);
     * schema.parse("да"); // true
     */
    addTruthyValues(values) {
        values.forEach((val) => this.truthyValues.add(val));
        return this;
    }
    /**
     * Adds custom falsy values for coercion.
     * These values will be converted to false when coerce() is enabled.
     *
     * @public
     * @param values - Array of values to treat as falsy
     * @returns this for method chaining
     *
     * @example
     * const schema = new BooleanSchema()
     *   .coerce()
     *   .addFalsyValues(["нет", "iie", "non"]);
     * schema.parse("нет"); // false
     */
    addFalsyValues(values) {
        values.forEach((val) => this.falsyValues.add(val));
        return this;
    }
}
exports.BooleanSchema = BooleanSchema;
