// validor/dist/schemas/DateSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DateSchema = void 0;
const BaseSchema_1 = require("../core/BaseSchema");
const ValidationError_1 = require("../errors/ValidationError");
const messages_1 = require("../i18n/messages");
class DateSchema extends BaseSchema_1.BaseSchema {
    constructor() {
        super(...arguments);
        this.checks = [];
    }
    parse(value, path = []) {
        // Handle null/undefined with defaults
        const result = this.handleNullUndefined(value, path);
        if (result.handled) {
            return result.value;
        }
        // Parse to Date
        const date = value instanceof Date ? value : new Date(value);
        // Validate date
        if (!(value instanceof Date ||
            typeof value === "string" ||
            typeof value === "number") ||
            isNaN(date.getTime())) {
            this.throwTypeError(path, "valid date");
        }
        // Run all checks
        this.runChecks(date, path);
        return this.transformFn ? this.transformFn(date) : date;
    }
    runChecks(value, path) {
        const issues = this.checks
            .filter(({ test }) => !test(value))
            .map(({ code, message, params }) => ({
            path,
            message: message || (0, messages_1.getMessage)(code, path, params),
            code,
        }));
        if (issues.length > 0) {
            throw new ValidationError_1.ValidationError(issues);
        }
    }
    addCheck(test, code, message, params) {
        this.checks.push({ test, code, message, params });
        return this;
    }
    min(date, message) {
        const minDate = new Date(date);
        return this.addCheck((val) => val >= minDate, "date_too_small", message, {
            min: minDate.toISOString(),
        });
    }
    max(date, message) {
        const maxDate = new Date(date);
        return this.addCheck((val) => val <= maxDate, "date_too_big", message, {
            max: maxDate.toISOString(),
        });
    }
    past(message) {
        return this.addCheck((val) => val < new Date(), "date_not_past", message);
    }
    future(message) {
        return this.addCheck((val) => val > new Date(), "date_not_future", message);
    }
    between(start, end, message) {
        const startDate = new Date(start);
        const endDate = new Date(end);
        return this.addCheck((val) => val >= startDate && val <= endDate, "date_out_of_range", message, { start: startDate.toISOString(), end: endDate.toISOString() });
    }
    // Useful transforms
    toISOString() {
        return this.transform((val) => val.toISOString());
    }
    toTimestamp() {
        return this.transform((val) => val.getTime());
    }
}
exports.DateSchema = DateSchema;
