// validor/dist/schemas/IPv4Schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IPv4Schema = void 0;
const BaseSchema_1 = require("../core/BaseSchema");
const ValidationError_1 = require("../errors/ValidationError");
const messages_1 = require("../i18n/messages");
/**
 * Schema class for validating IPv4 addresses.
 * Supports validation of standard IPv4 format, CIDR notation, and range checks.
 *
 * @example
 * const schema = new IPv4Schema();
 * schema.parse("192.168.1.1");    // "192.168.1.1"
 * schema.parse("256.1.1.1");      // throws ValidationError
 *
 * @example
 * // With CIDR notation
 * const schema = new IPv4Schema().cidr();
 * schema.parse("192.168.1.0/24"); // "192.168.1.0/24"
 */
class IPv4Schema extends BaseSchema_1.BaseSchema {
    constructor() {
        super(...arguments);
        /**
         * List of validation checks to run against the value.
         * Each check contains a test function, error code, and optional custom message.
         * @private
         */
        this.checks = [];
        /**
         * Flag indicating whether CIDR notation is allowed.
         * @private
         */
        this.allowCidr = false;
        /**
         * Flag indicating whether CIDR notation is required.
         * @private
         */
        this.requireCidr = false;
        /**
         * Regular expression pattern for validating IPv4 addresses.
         * Matches four octets (0-255) separated by dots.
         * @private
         */
        this.ipv4Pattern = /^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/;
        /**
         * Regular expression pattern for validating IPv4 addresses with CIDR notation.
         * Matches IPv4 address followed by /0-32.
         * @private
         */
        this.cidrPattern = /^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\/(?:3[0-2]|[12]?[0-9])$/;
    }
    /**
     * Parses and validates the given value as an IPv4 address.
     *
     * @public
     * @param value - The value to validate
     * @param path - The field path for error messages
     * @returns The validated IPv4 address string
     * @throws {ValidationError} If validation fails
     *
     * @example
     * const schema = new IPv4Schema();
     * schema.parse("192.168.1.1"); // "192.168.1.1"
     * schema.parse("invalid");     // throws ValidationError
     */
    parse(value, path = []) {
        // Handle null/undefined with defaults
        const result = this.handleNullUndefined(value, path);
        if (result.handled) {
            return result.value;
        }
        // Type check
        if (typeof value !== "string") {
            this.throwTypeError(path, "string");
        }
        const trimmedValue = value.trim();
        // Validate IPv4 format
        if (!this.isValidIPv4(trimmedValue)) {
            throw new ValidationError_1.ValidationError([
                {
                    path,
                    message: (0, messages_1.getMessage)("invalid_ipv4", path),
                    code: "invalid_ipv4",
                },
            ]);
        }
        // Run all additional checks
        this.runChecks(trimmedValue, path);
        // Apply transform if exists
        return this.transformFn ? this.transformFn(trimmedValue) : trimmedValue;
    }
    runChecks(value, path) {
        const issues = this.checks
            .filter(({ test }) => !test(value))
            .map(({ code, message, params }) => ({
            path,
            message: message || (0, messages_1.getMessage)(code, path, params),
            code,
        }));
        if (issues.length > 0) {
            throw new ValidationError_1.ValidationError(issues);
        }
    }
    /**
     * Validates whether the given string is a valid IPv4 address.
     * Handles both standard format and CIDR notation based on schema configuration.
     *
     * @private
     * @param value - The string to validate
     * @returns true if valid, false otherwise
     */
    isValidIPv4(value) {
        const hasCidr = value.includes("/");
        if (this.requireCidr) {
            return this.cidrPattern.test(value);
        }
        if (hasCidr) {
            return this.allowCidr && this.cidrPattern.test(value);
        }
        return this.ipv4Pattern.test(value);
    }
    /**
     * Adds a new validation check to the schema.
     *
     * @private
     * @param test - The test function that returns true if validation passes
     * @param code - The error code for i18n message lookup
     * @param message - Optional custom error message
     * @param params - Optional parameters to include in error message
     * @returns this for method chaining
     */
    addCheck(test, code, message, params) {
        this.checks.push({ test, code, message, params });
        return this;
    }
    /**
     * Converts an IPv4 address string to a numeric value for comparison.
     *
     * @private
     * @param ip - The IPv4 address string
     * @returns The numeric representation of the IP address
     *
     * @example
     * this.ipToNumber("192.168.1.1"); // 3232235777
     */
    ipToNumber(ip) {
        const ipOnly = ip.split("/")[0];
        const parts = ipOnly.split(".").map(Number);
        return (((parts[0] << 24) | (parts[1] << 16) | (parts[2] << 8) | parts[3]) >>> 0);
    }
    /**
     * Extracts the IP address part from a value that may contain CIDR notation.
     *
     * @private
     * @param value - The IPv4 address string, possibly with CIDR notation
     * @returns The IP address without CIDR suffix
     */
    getIpPart(value) {
        return value.split("/")[0];
    }
    /**
     * Gets the IP octets as an array of numbers.
     *
     * @private
     * @param value - The IPv4 address string, possibly with CIDR notation
     * @returns Array of four octets
     */
    getIpParts(value) {
        return this.getIpPart(value).split(".").map(Number);
    }
    /**
     * Checks if an IP is in any of the non-public ranges.
     *
     * @private
     * @param parts - Array of four octets
     * @returns true if IP is public, false if private/reserved
     */
    isPublicIP(parts) {
        // Private ranges
        if (parts[0] === 10)
            return false;
        if (parts[0] === 172 && parts[1] >= 16 && parts[1] <= 31)
            return false;
        if (parts[0] === 192 && parts[1] === 168)
            return false;
        // Special ranges
        if (parts[0] === 0)
            return false; // 0.0.0.0/8
        if (parts[0] === 127)
            return false; // Loopback 127.0.0.0/8
        if (parts[0] === 169 && parts[1] === 254)
            return false; // Link-local 169.254.0.0/16
        if (parts[0] >= 224 && parts[0] <= 239)
            return false; // Multicast 224.0.0.0/4
        if (parts[0] >= 240)
            return false; // Reserved 240.0.0.0/4
        return true;
    }
    /**
     * Allows CIDR notation in the IPv4 address (e.g., "192.168.1.0/24").
     *
     * @public
     * @returns this for method chaining
     *
     * @example
     * const schema = new IPv4Schema().cidr();
     * schema.parse("192.168.1.0/24"); // "192.168.1.0/24"
     * schema.parse("192.168.1.1");    // "192.168.1.1" (still valid)
     */
    cidr() {
        this.allowCidr = true;
        return this;
    }
    /**
     * Requires CIDR notation in the IPv4 address.
     * The address must include a subnet mask (e.g., "192.168.1.0/24").
     *
     * @public
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * const schema = new IPv4Schema().cidrRequired();
     * schema.parse("192.168.1.0/24"); // "192.168.1.0/24"
     * schema.parse("192.168.1.1");    // throws ValidationError
     */
    cidrRequired(message) {
        this.allowCidr = true;
        this.requireCidr = true;
        return this;
    }
    /**
     * Validates that the IP address is within a private network range.
     * Private ranges: 10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16
     *
     * @public
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * const schema = new IPv4Schema().private();
     * schema.parse("192.168.1.1"); // "192.168.1.1"
     * schema.parse("8.8.8.8");     // throws ValidationError
     */
    private(message) {
        return this.addCheck((val) => {
            const parts = this.getIpParts(val);
            // 10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16
            if (parts[0] === 10)
                return true;
            if (parts[0] === 172 && parts[1] >= 16 && parts[1] <= 31)
                return true;
            if (parts[0] === 192 && parts[1] === 168)
                return true;
            return false;
        }, "ipv4_not_private", message);
    }
    /**
     * Validates that the IP address is a public (non-private, non-reserved) address.
     *
     * @public
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * const schema = new IPv4Schema().public();
     * schema.parse("8.8.8.8");     // "8.8.8.8"
     * schema.parse("192.168.1.1"); // throws ValidationError
     */
    public(message) {
        return this.addCheck((val) => this.isPublicIP(this.getIpParts(val)), "ipv4_not_public", message);
    }
    /**
     * Validates that the IP address is a loopback address (127.0.0.0/8).
     *
     * @public
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * const schema = new IPv4Schema().loopback();
     * schema.parse("127.0.0.1"); // "127.0.0.1"
     * schema.parse("127.0.1.1"); // "127.0.1.1"
     * schema.parse("8.8.8.8");   // throws ValidationError
     */
    loopback(message) {
        return this.addCheck((val) => this.getIpParts(val)[0] === 127, "ipv4_not_loopback", message);
    }
    /**
     * Validates that the IP address is NOT a loopback address.
     *
     * @public
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * const schema = new IPv4Schema().notLoopback();
     * schema.parse("192.168.1.1"); // "192.168.1.1"
     * schema.parse("127.0.0.1");   // throws ValidationError
     */
    notLoopback(message) {
        return this.addCheck((val) => this.getIpParts(val)[0] !== 127, "ipv4_is_loopback", message);
    }
    /**
     * Validates that the IP address is within a specified range.
     *
     * @public
     * @param startIp - The start of the IP range (inclusive)
     * @param endIp - The end of the IP range (inclusive)
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * const schema = new IPv4Schema().range("192.168.1.1", "192.168.1.254");
     * schema.parse("192.168.1.100"); // "192.168.1.100"
     * schema.parse("192.168.2.1");   // throws ValidationError
     */
    range(startIp, endIp, message) {
        const startNum = this.ipToNumber(startIp);
        const endNum = this.ipToNumber(endIp);
        return this.addCheck((val) => {
            const ip = this.getIpPart(val);
            const num = this.ipToNumber(ip);
            return num >= startNum && num <= endNum;
        }, "ipv4_out_of_range", message, { start: startIp, end: endIp });
    }
    /**
     * Validates that the IP address is within a specified subnet.
     *
     * @public
     * @param subnet - The subnet in CIDR notation (e.g., "192.168.1.0/24")
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * const schema = new IPv4Schema().inSubnet("192.168.1.0/24");
     * schema.parse("192.168.1.100"); // "192.168.1.100"
     * schema.parse("192.168.2.1");   // throws ValidationError
     */
    inSubnet(subnet, message) {
        const [subnetIp, prefixStr] = subnet.split("/");
        const prefix = parseInt(prefixStr, 10);
        const mask = prefix === 0 ? 0 : (~0 << (32 - prefix)) >>> 0;
        const subnetNum = this.ipToNumber(subnetIp) & mask;
        return this.addCheck((val) => {
            const ip = this.getIpPart(val);
            const ipNum = this.ipToNumber(ip);
            return (ipNum & mask) === subnetNum;
        }, "ipv4_not_in_subnet", message, { subnet });
    }
    /**
     * Validates that the IP address is NOT within a specified subnet.
     *
     * @public
     * @param subnet - The subnet in CIDR notation (e.g., "10.0.0.0/8")
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * const schema = new IPv4Schema().notInSubnet("10.0.0.0/8");
     * schema.parse("192.168.1.1"); // "192.168.1.1"
     * schema.parse("10.0.0.1");    // throws ValidationError
     */
    notInSubnet(subnet, message) {
        const [subnetIp, prefixStr] = subnet.split("/");
        const prefix = parseInt(prefixStr, 10);
        const mask = prefix === 0 ? 0 : (~0 << (32 - prefix)) >>> 0;
        const subnetNum = this.ipToNumber(subnetIp) & mask;
        return this.addCheck((val) => {
            const ip = this.getIpPart(val);
            const ipNum = this.ipToNumber(ip);
            return (ipNum & mask) !== subnetNum;
        }, "ipv4_in_blocked_subnet", message, { subnet });
    }
    /**
     * Validates the CIDR prefix length is within a specified range.
     * Only applicable when CIDR notation is used.
     *
     * @public
     * @param min - Minimum prefix length (0-32)
     * @param max - Maximum prefix length (0-32)
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * const schema = new IPv4Schema().cidr().prefixLength(24, 30);
     * schema.parse("192.168.1.0/24"); // "192.168.1.0/24"
     * schema.parse("192.168.1.0/16"); // throws ValidationError
     */
    prefixLength(min, max, message) {
        return this.addCheck((val) => {
            if (!val.includes("/"))
                return true; // Skip if no CIDR
            const prefix = parseInt(val.split("/")[1], 10);
            return prefix >= min && prefix <= max;
        }, "ipv4_invalid_prefix", message, { min, max });
    }
    /**
     * Validates that the IP address is a valid network address for its CIDR prefix.
     * The host bits must all be zero.
     *
     * @public
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * const schema = new IPv4Schema().cidr().networkAddress();
     * schema.parse("192.168.1.0/24");   // "192.168.1.0/24"
     * schema.parse("192.168.1.100/24"); // throws ValidationError
     */
    networkAddress(message) {
        return this.addCheck((val) => {
            if (!val.includes("/"))
                return true;
            const [ip, prefixStr] = val.split("/");
            const prefix = parseInt(prefixStr, 10);
            const mask = prefix === 0 ? 0 : (~0 << (32 - prefix)) >>> 0;
            const ipNum = this.ipToNumber(ip);
            const maskedIp = (ipNum & mask) >>> 0;
            return maskedIp === ipNum;
        }, "ipv4_not_network_address", message);
    }
    /**
     * Blocks reserved IP addresses (0.0.0.0/8, 240.0.0.0/4, etc.).
     *
     * @public
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * const schema = new IPv4Schema().notReserved();
     * schema.parse("192.168.1.1"); // "192.168.1.1"
     * schema.parse("0.0.0.1");     // throws ValidationError
     * schema.parse("255.255.255.255"); // throws ValidationError
     */
    notReserved(message) {
        return this.addCheck((val) => {
            const ip = this.getIpPart(val);
            const parts = ip.split(".").map(Number);
            // 0.0.0.0/8
            if (parts[0] === 0)
                return false;
            // 240.0.0.0/4 (reserved for future use)
            if (parts[0] >= 240)
                return false;
            return true;
        }, "ipv4_reserved", message);
    }
}
exports.IPv4Schema = IPv4Schema;
