// validor/dist/schemas/NumberSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NumberSchema = void 0;
const BaseSchema_1 = require("../core/BaseSchema");
const ValidationError_1 = require("../errors/ValidationError");
const messages_1 = require("../i18n/messages");
class NumberSchema extends BaseSchema_1.BaseSchema {
    constructor() {
        super(...arguments);
        this.checks = [];
    }
    parse(value, path = []) {
        // Handle null/undefined with defaults
        const result = this.handleNullUndefined(value, path);
        if (result.handled) {
            return result.value;
        }
        // Type check
        if (typeof value !== "number" || isNaN(value)) {
            this.throwTypeError(path, "number");
        }
        // Run all checks
        this.runChecks(value, path);
        return this.transformFn ? this.transformFn(value) : value;
    }
    runChecks(value, path) {
        const issues = this.checks
            .filter(({ test }) => !test(value))
            .map(({ code, message, params }) => ({
            path,
            message: message || (0, messages_1.getMessage)(code, path, params),
            code,
        }));
        if (issues.length > 0) {
            throw new ValidationError_1.ValidationError(issues);
        }
    }
    addCheck(test, code, message, params) {
        this.checks.push({ test, code, message, params });
        return this;
    }
    min(n, message) {
        return this.addCheck((val) => val >= n, "too_small", message, { min: n });
    }
    max(n, message) {
        return this.addCheck((val) => val <= n, "too_big", message, { max: n });
    }
    positive(message) {
        return this.addCheck((val) => val > 0, "not_positive", message);
    }
    negative(message) {
        return this.addCheck((val) => val < 0, "not_negative", message);
    }
    nonnegative(message) {
        return this.addCheck((val) => val >= 0, "negative", message);
    }
    integer(message) {
        return this.addCheck((val) => Number.isInteger(val), "not_integer", message);
    }
    finite(message) {
        return this.addCheck((val) => Number.isFinite(val), "not_finite", message);
    }
    multipleOf(n, message) {
        return this.addCheck((val) => {
            // Check if val/n is close to an integer for better floating point handling
            const quotient = val / n;
            const rounded = Math.round(quotient);
            const epsilon = 1e-10; // Tolerance for floating point errors
            return Math.abs(quotient - rounded) < epsilon;
        }, "not_multiple", message, { multiple: n });
    }
    between(min, max, message) {
        return this.addCheck((val) => val >= min && val <= max, "out_of_range", message, { min, max });
    }
}
exports.NumberSchema = NumberSchema;
