// validor/dist/schemas/ObjectSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectSchema = void 0;
const BaseSchema_1 = require("../core/BaseSchema");
const ValidationError_1 = require("../errors/ValidationError");
const messages_1 = require("../i18n/messages");
class ObjectSchema extends BaseSchema_1.BaseSchema {
    constructor(shape) {
        super();
        this.shape = shape;
        this.strictMode = false;
    }
    parse(value, path = []) {
        // Apply default if value is undefined
        if (value === undefined) {
            const defaultVal = this.getDefaultValue();
            if (defaultVal !== undefined) {
                value = defaultVal;
            }
        }
        const checked = this.checkValue(value, path);
        if (checked === null)
            return value;
        const result = this.validateFieldsSync(checked, path);
        return this.transformFn ? this.transformFn(result) : result;
    }
    async validate(value, path = []) {
        this.path = path;
        // Apply default if value is undefined
        if (value === undefined) {
            const defaultVal = this.getDefaultValue();
            if (defaultVal !== undefined) {
                value = defaultVal;
            }
        }
        const checked = this.checkValue(value, path);
        if (checked === null)
            return value;
        return this.validateFieldsAsync(checked, path);
    }
    // Returns validated object or null (for null/undefined that should pass through)
    checkValue(value, path) {
        if (value == null) {
            // Handle null specifically
            if (value === null) {
                if (this.isNullable)
                    return null;
                throw new ValidationError_1.ValidationError([
                    {
                        path,
                        message: (0, messages_1.getMessage)("invalid_type", path, { expected: "object" }),
                        code: "invalid_type",
                    },
                ]);
            }
            // Handle undefined
            if (this.isRequiredFlag) {
                throw new ValidationError_1.ValidationError([
                    {
                        path,
                        message: this.requiredMessage || (0, messages_1.getMessage)("required", path),
                        code: "required",
                    },
                ]);
            }
            return null;
        }
        if (typeof value !== "object" || Array.isArray(value)) {
            throw new ValidationError_1.ValidationError([
                {
                    path,
                    message: (0, messages_1.getMessage)("invalid_type", path, { expected: "object" }),
                    code: "invalid_type",
                },
            ]);
        }
        // Strict mode: check for unknown keys
        if (this.strictMode) {
            const unknownKeys = Object.keys(value).filter((k) => !(k in this.shape));
            if (unknownKeys.length > 0) {
                throw new ValidationError_1.ValidationError(unknownKeys.map((key) => ({
                    path: [...path, key],
                    message: (0, messages_1.getMessage)("unknown_key", [...path, key]),
                    code: "unknown_key",
                })));
            }
        }
        return value;
    }
    validateFieldsSync(value, path) {
        const result = {};
        const issues = [];
        for (const [key, schema] of Object.entries(this.shape)) {
            const fieldPath = [...path, key];
            try {
                result[key] = schema.parse(value[key], fieldPath);
            }
            catch (err) {
                this.collectError(err, fieldPath, issues);
            }
        }
        if (issues.length > 0)
            throw new ValidationError_1.ValidationError(issues);
        return result;
    }
    async validateFieldsAsync(value, path) {
        const result = {};
        const issues = [];
        // Run all validations in parallel
        const entries = Object.entries(this.shape);
        const results = await Promise.allSettled(entries.map(async ([key, schema]) => ({
            key,
            value: await schema.validate(value[key], [...path, key]),
        })));
        for (let i = 0; i < results.length; i++) {
            const res = results[i];
            const key = entries[i][0];
            if (res.status === "fulfilled") {
                result[key] = res.value.value;
            }
            else {
                this.collectError(res.reason, [...path, key], issues);
            }
        }
        if (issues.length > 0)
            throw new ValidationError_1.ValidationError(issues);
        const finalResult = this.transformFn ? this.transformFn(result) : result;
        return finalResult;
    }
    collectError(err, fieldPath, issues) {
        if (err instanceof ValidationError_1.ValidationError) {
            issues.push(...err.issues);
        }
        else {
            issues.push({
                path: fieldPath,
                message: err instanceof Error ? err.message : "Validation failed",
                code: "invalid",
            });
        }
    }
    // Utility methods
    strict() {
        this.strictMode = true;
        return this;
    }
    passthrough() {
        this.strictMode = false;
        return this;
    }
    extend(shape) {
        return new ObjectSchema({ ...this.shape, ...shape });
    }
    pick(...keys) {
        const picked = keys.reduce((acc, key) => {
            acc[key] = this.shape[key];
            return acc;
        }, {});
        return new ObjectSchema(picked);
    }
    omit(...keys) {
        const omitted = Object.fromEntries(Object.entries(this.shape).filter(([k]) => !keys.includes(k)));
        return new ObjectSchema(omitted);
    }
    partial() {
        const partial = Object.fromEntries(Object.entries(this.shape).map(([k, v]) => [k, v.optional()]));
        return new ObjectSchema(partial);
    }
    merge(other) {
        return new ObjectSchema({ ...this.shape, ...other.shape });
    }
}
exports.ObjectSchema = ObjectSchema;
