// validor/dist/schemas/StringSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StringSchema = void 0;
const BaseSchema_1 = require("../core/BaseSchema");
const ValidationError_1 = require("../errors/ValidationError");
const messages_1 = require("../i18n/messages");
/**
 * Schema class for validating string values.
 * Supports chainable validation rules and transformations.
 *
 * @example
 * const schema = new StringSchema()
 *   .trim()
 *   .min(3)
 *   .max(50)
 *   .email();
 */
class StringSchema extends BaseSchema_1.BaseSchema {
    constructor() {
        super(...arguments);
        /**
         * List of validation checks to run against the value.
         * Each check contains a test function, error code, and optional custom message.
         * @private
         */
        this.checks = [];
        /**
         * List of transformation functions to apply before validation.
         * Transforms are applied in order they were added.
         * @private
         */
        this.transforms = [];
    }
    /**
     * Parses and validates the given value as a string.
     *
     * @public
     * @param value - The value to validate
     * @param path - The field path for error messages
     * @returns The transformed and validated string
     * @throws {ValidationError} If validation fails
     *
     * @example
     * const schema = new StringSchema().min(3);
     * schema.parse("hello"); // "hello"
     * schema.parse("hi");    // throws ValidationError
     */
    parse(value, path = []) {
        // Handle null/undefined with defaults
        const result = this.handleNullUndefined(value, path);
        if (result.handled) {
            return result.value;
        }
        // Type check
        if (typeof value !== "string") {
            this.throwTypeError(path, "string");
        }
        // Apply string-specific transforms first (before validation)
        let transformedValue = value;
        for (const transform of this.transforms) {
            transformedValue = transform(transformedValue);
        }
        // Run all checks on transformed value
        this.runChecks(transformedValue, path);
        return this.transformFn ? this.transformFn(transformedValue) : transformedValue;
    }
    /**
     * Runs all validation checks and throws if any fail
     * @private
     */
    runChecks(value, path) {
        const issues = this.checks
            .filter(({ test }) => !test(value))
            .map(({ code, message, params }) => ({
            path,
            message: message || (0, messages_1.getMessage)(code, path, params),
            code,
        }));
        if (issues.length > 0) {
            throw new ValidationError_1.ValidationError(issues);
        }
    }
    /**
     * Adds a new validation check to the schema.
     *
     * @private
     * @param test - The test function that returns true if validation passes
     * @param code - The error code for i18n message lookup
     * @param message - Optional custom error message
     * @param params - Optional parameters to include in error message
     * @returns this for method chaining
     */
    addCheck(test, code, message, params) {
        this.checks.push({ test, code, message, params });
        return this;
    }
    /**
     * Adds a minimum length validation check.
     *
     * @public
     * @param n - The minimum number of characters required
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * new StringSchema().min(3).parse("ab"); // throws ValidationError
     * new StringSchema().min(3).parse("abc"); // "abc"
     */
    min(n, message) {
        return this.addCheck((val) => val.length >= n, "too_small", message, {
            min: n,
        });
    }
    /**
     * Adds a maximum length validation check.
     *
     * @public
     * @param n - The maximum number of characters allowed
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * new StringSchema().max(5).parse("hello"); // "hello"
     * new StringSchema().max(5).parse("hello world"); // throws ValidationError
     */
    max(n, message) {
        return this.addCheck((val) => val.length <= n, "too_big", message, {
            max: n,
        });
    }
    /**
     * Adds an exact length validation check.
     *
     * @public
     * @param n - The exact number of characters required
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * new StringSchema().length(5).parse("hello"); // "hello"
     * new StringSchema().length(5).parse("hi");    // throws ValidationError
     */
    length(n, message) {
        return this.addCheck((val) => val.length === n, "invalid_length", message, {
            length: n,
        });
    }
    /**
     * Adds an email format validation check.
     * Uses a basic regex pattern for email validation.
     *
     * @public
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * new StringSchema().email().parse("test@example.com"); // "test@example.com"
     * new StringSchema().email().parse("invalid-email");    // throws ValidationError
     */
    email(message) {
        return this.addCheck((val) => /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(val), "invalid_email", message);
    }
    /**
     * Adds a URL format validation check.
     * Uses the native URL constructor for validation.
     *
     * @public
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * new StringSchema().url().parse("https://example.com"); // "https://example.com"
     * new StringSchema().url().parse("not-a-url");           // throws ValidationError
     */
    url(message) {
        return this.addCheck((val) => {
            try {
                new URL(val);
                return true;
            }
            catch {
                return false;
            }
        }, "invalid_url", message);
    }
    /**
     * Adds a regex pattern validation check.
     *
     * @public
     * @param pattern - The RegExp pattern to test against
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * new StringSchema().regex(/^[A-Z]+$/).parse("HELLO"); // "HELLO"
     * new StringSchema().regex(/^[A-Z]+$/).parse("hello"); // throws ValidationError
     */
    regex(pattern, message) {
        return this.addCheck((val) => pattern.test(val), "invalid_regex", message);
    }
    /**
     * Adds a non-empty string validation check.
     * Ensures the string has at least one character.
     *
     * @public
     * @param message - Optional custom error message
     * @returns this for method chaining
     *
     * @example
     * new StringSchema().nonempty().parse("hello"); // "hello"
     * new StringSchema().nonempty().parse("");      // throws ValidationError
     */
    nonempty(message) {
        return this.addCheck((val) => val.length > 0, "empty_string", message);
    }
    /**
     * Adds a trim transformation that removes leading and trailing whitespace.
     * Applied before validation checks.
     *
     * @public
     * @returns this for method chaining
     *
     * @example
     * new StringSchema().trim().parse("  hello  "); // "hello"
     */
    trim() {
        this.transforms.push((val) => val.trim());
        return this;
    }
    /**
     * Adds a lowercase transformation that converts all characters to lowercase.
     * Applied before validation checks.
     *
     * @public
     * @returns this for method chaining
     *
     * @example
     * new StringSchema().lowercase().parse("HELLO"); // "hello"
     */
    lowercase() {
        this.transforms.push((val) => val.toLowerCase());
        return this;
    }
    /**
     * Adds an uppercase transformation that converts all characters to uppercase.
     * Applied before validation checks.
     *
     * @public
     * @returns this for method chaining
     *
     * @example
     * new StringSchema().uppercase().parse("hello"); // "HELLO"
     */
    uppercase() {
        this.transforms.push((val) => val.toUpperCase());
        return this;
    }
}
exports.StringSchema = StringSchema;
