// validor/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.i18n = exports.getLocale = exports.setLocale = exports.array = exports.IPv4 = exports.boolean = exports.object = exports.date = exports.number = exports.string = exports.ValidationError = exports.BaseSchema = void 0;
var BaseSchema_1 = require("./core/BaseSchema");
Object.defineProperty(exports, "BaseSchema", { enumerable: true, get: function () { return BaseSchema_1.BaseSchema; } });
var ValidationError_1 = require("./errors/ValidationError");
Object.defineProperty(exports, "ValidationError", { enumerable: true, get: function () { return ValidationError_1.ValidationError; } });
// Validators
var string_1 = require("./validators/string");
Object.defineProperty(exports, "string", { enumerable: true, get: function () { return string_1.string; } });
var number_1 = require("./validators/number");
Object.defineProperty(exports, "number", { enumerable: true, get: function () { return number_1.number; } });
var date_1 = require("./validators/date");
Object.defineProperty(exports, "date", { enumerable: true, get: function () { return date_1.date; } });
var object_1 = require("./validators/object");
Object.defineProperty(exports, "object", { enumerable: true, get: function () { return object_1.object; } });
var boolean_1 = require("./validators/boolean");
Object.defineProperty(exports, "boolean", { enumerable: true, get: function () { return boolean_1.boolean; } });
var IPv4_1 = require("./validators/IPv4");
Object.defineProperty(exports, "IPv4", { enumerable: true, get: function () { return IPv4_1.IPv4; } });
var array_1 = require("./validators/array");
Object.defineProperty(exports, "array", { enumerable: true, get: function () { return array_1.array; } });
// Utils
var messages_1 = require("./i18n/messages");
Object.defineProperty(exports, "setLocale", { enumerable: true, get: function () { return messages_1.setLocale; } });
Object.defineProperty(exports, "getLocale", { enumerable: true, get: function () { return messages_1.getLocale; } });
var i18n_1 = require("./i18n/i18n");
Object.defineProperty(exports, "i18n", { enumerable: true, get: function () { return i18n_1.i18n; } });
// Set default locale to English
const messages_2 = require("./i18n/messages");
(0, messages_2.setLocale)("en");
